### DISCUSSION TITLE
MIgrating form helm chart to operator - questions

### DISCUSSION THREAD

**Discussion by grzesuav** [cryostatio/cryostat#745] [General]
Hi, I am trying to migrate from chelm chart to use operator, I have cryostat cr 
```yaml
apiVersion: operator.cryostat.io/v1beta2
kind: Cryostat
metadata:
  annotations:
    argocd.argoproj.io/tracking-id: >-
      argocd-compute_cryostat-cr-staging:operator.cryostat.io/Cryostat:compute-platform-observability/cryostat-main
  name: cryostat-main
  namespace: compute-platform-observability
spec:
  enableCertManager: true
  eventTemplates: []
  operandMetadata:
    podMetadata:
      annotations:
        sidecar.istio.io/inject: 'false'
  reportOptions:
    replicas: 0
  resources:
    coreResources:
      requests:
        memory: 1Gi
    dataSourceResources:
      requests:
        memory: 700Mi
...
  serviceOptions:
    coreConfig:
      httpPort: 8181
      serviceType: ClusterIP
  storageOptions:
    pvc:
      annotations: {}
      labels: {}
      spec:
        accessModes:
          - ReadWriteOnce
        resources:
          requests:
            storage: 5Gi
        storageClassName: managed-csi
  targetDiscoveryOptions:
    discoveryPortNames:
      - tcp-jfr-jmx
    discoveryPortNumbers:
      - 9091
  targetNamespaces:
...
  trustedCertSecrets: []

```
however I see some differences with the target instance. I do not want cryostat to expose https, I want only to expose http, however with the following it creates

```yaml
- args:
    - --alpha-config=/etc/oauth2_proxy/alpha_config/alpha_config.json
    env:
    - name: OAUTH2_PROXY_REDIRECT_URL
      value: http://localhost:4180/oauth2/callback
    - name: OAUTH2_PROXY_EMAIL_DOMAINS
      value: '*'
    - name: OAUTH2_PROXY_SKIP_AUTH_ROUTES
      value: .*
    envFrom:
    - secretRef:
        name: cryostat-main-oauth2-cookie
        optional: false
    image: quay.io/oauth2-proxy/oauth2-proxy:latest
    imagePullPolicy: Always
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /ping
        port: 4180
        scheme: HTTPS
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    name: cryostat-main-auth-proxy
    ports:
    - containerPort: 4180
      protocol: TCP
...
```

when I do 
```
k port-forward svc/cryostat-main  8181:8181
```
I see the port on cryostat is exposed as https (I have waring in the browser that it got http request on https).

What I need to set on CR to make it start on http ?

**Comment by grzesuav** [comment]
ok, im operator setup the proxy image is set to `image: quay.io/oauth2-proxy/oauth2-proxy:latest`, in helm chart it is set to fixed version, will try to adjust that - to use `quay.io/oauth2-proxy/oauth2-proxy:v7.6.0`

**Comment by andrewazores** [comment]
https://github.com/cryostatio/cryostat-operator?tab=readme-ov-file#security

Can you try disabling the cert-manager integration?

**Reply by grzesuav** [reply]
When I set it to false, the container is crashlooping - 
```
❯ k logs cryostat-main-556654f588-gr5q6 -c cryostat-main-auth-proxy
[2024/12/16 16:00:54] [main.go:71] WARNING: You are using alpha configuration. The structure in this configuration file may change without notice. You MUST remove conflicting options from your existing configuration.
[2024/12/16 16:00:54] [proxy.go:89] mapping path "^/storage/(.*)$" => upstream "http://localhost:8333"
[2024/12/16 16:00:54] [proxy.go:89] mapping path "/grafana/" => upstream "http://localhost:3000"
[2024/12/16 16:00:54] [proxy.go:89] mapping path "/" => upstream "http://localhost:8181"
[2024/12/16 16:00:54] [oauthproxy.go:171] OAuthProxy configured for Google Client ID: CLIENT_ID
[2024/12/16 16:00:54] [oauthproxy.go:177] Cookie settings: name:_oauth2_proxy secure(https):true httponly:true expiry:168h0m0s domains: path:/ samesite: refresh:disabled
[2024/12/16 16:00:54] [oauthproxy.go:498] Skipping auth - Method:  | Path: .*
[2024/12/16 16:00:54] [main.go:58] ERROR: Failed to initialise OAuth2 Proxy: error setting up server: could not build app server: error setting up TLS listener: could not load certificate: could not load key data: open /var/run/secrets/operator.cryostat.io/cryostat-main-tls/tls.key: no such file or directory
❯ k logs cryostat-main-556654f588-gr5q6 -c cryostat-main-auth-proxy --previous
[2024/12/16 16:00:54] [main.go:71] WARNING: You are using alpha configuration. The structure in this configuration file may change without notice. You MUST remove conflicting options from your existing configuration.
[2024/12/16 16:00:54] [proxy.go:89] mapping path "^/storage/(.*)$" => upstream "http://localhost:8333"
[2024/12/16 16:00:54] [proxy.go:89] mapping path "/grafana/" => upstream "http://localhost:3000"
[2024/12/16 16:00:54] [proxy.go:89] mapping path "/" => upstream "http://localhost:8181"
[2024/12/16 16:00:54] [oauthproxy.go:171] OAuthProxy configured for Google Client ID: CLIENT_ID
[2024/12/16 16:00:54] [oauthproxy.go:177] Cookie settings: name:_oauth2_proxy secure(https):true httponly:true expiry:168h0m0s domains: path:/ samesite: refresh:disabled
[2024/12/16 16:00:54] [oauthproxy.go:498] Skipping auth - Method:  | Path: .*
[2024/12/16 16:00:54] [main.go:58] ERROR: Failed to initialise OAuth2 Proxy: error setting up server: could not build app server: error setting up TLS listener: could not load certificate: could not load key data: open /var/run/secrets/operator.cryostat.io/cryostat-main-tls/tls.key: no such file or directory
  │  ~ 
```

for record 
```yaml
apiVersion: operator.cryostat.io/v1beta2
kind: Cryostat
metadata:
  name: cryostat-main
  namespace: compute-platform-observability
spec:
  resources:
    coreResources:
      requests:
        memory: 1Gi
    dataSourceResources:
      requests:
        memory: 700Mi
  serviceOptions:
    coreConfig:
      httpPort: 8181
      serviceType: ClusterIP
  targetNamespaces:
...
  targetDiscoveryOptions:
    discoveryPortNames:
      - tcp-jfr-jmx
    discoveryPortNumbers:
      - 9091
  enableCertManager: false
  trustedCertSecrets: []
  eventTemplates: []
  schedulingOptions:
...
  operandMetadata:
    podMetadata:
      annotations:
        sidecar.istio.io/inject: "false"
...

  reportOptions:
    replicas: 0

```

**Reply by grzesuav** [reply]
the oauth-proxy container definition in pod 
```yaml
  - args:
    - --alpha-config=/etc/oauth2_proxy/alpha_config/alpha_config.json
    env:
    - name: OAUTH2_PROXY_REDIRECT_URL
      value: http://localhost:4180/oauth2/callback
    - name: OAUTH2_PROXY_EMAIL_DOMAINS
      value: '*'
    - name: OAUTH2_PROXY_SKIP_AUTH_ROUTES
      value: .*
    envFrom:
    - secretRef:
        name: cryostat-main-oauth2-cookie
        optional: false
    image: quay.io/oauth2-proxy/oauth2-proxy:v7.6.0
    imagePullPolicy: IfNotPresent
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /ping
        port: 4180
        scheme: HTTP
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 1
    name: cryostat-main-auth-proxy
    ports:
    - containerPort: 4180
      protocol: TCP
    resources:
      requests:
        cpu: 25m
        memory: 64Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/oauth2_proxy/alpha_config
      name: cryostat-main-oauth2-proxy-cfg
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-f6fjh
      readOnly: true
```


so it is correctly recognizing 
```yaml     
     httpGet:
        path: /ping
        port: 4180
        scheme: HTTP
```
but unsure form where expecation about tls comes grom

**Reply by andrewazores** [reply]
Hmm, okay. That's an unfortunate bug. I filed it here: https://github.com/cryostatio/cryostat-operator/issues/984

**Reply by andrewazores** [reply]
https://github.com/cryostatio/cryostat-operator/blob/85d6445d2c1110e1b98c41a438d0cddac34e03a0/internal/controllers/common/resource_definitions/resource_definitions.go#L810

https://github.com/cryostatio/cryostat-operator/blob/85d6445d2c1110e1b98c41a438d0cddac34e03a0/internal/controllers/reconciler.go#L222

I'm not actually sure how/why that isn't working as expected. It looks like if TLS is disabled (ie cert-manager integration is disabled) then it should not try to mount the TLS secrets, but the crash log you shared shows that TLS secrets are the reason it's crashing.

**Reply by tthvo** [reply]
https://github.com/cryostatio/cryostat-operator/issues/984#issuecomment-2547657509

**Reply by grzesuav** [reply]
thanks, this one get solved. Another question, my instance from CR is missing grafana url, when I look at this code - https://github.com/cryostatio/cryostat-operator/blob/v3.0.1/internal/controllers/common/resource_definitions/resource_definitions.go#L1132 I am not sure how to drive this variable, as I do not set it in my CR. Any suggestion ?

**Reply by andrewazores** [reply]
Since 3.0, the Grafana dashboard container is normally deployed in the same Pod as Cryostat, hidden behind the same auth proxy instance. This way the two share the same authentication/authorization. Since you are not using OpenShift and Routes, this `spec.AuthProxyURL` should get set here:

https://github.com/cryostatio/cryostat-operator/blob/85d6445d2c1110e1b98c41a438d0cddac34e03a0/internal/controllers/ingresses.go#L51

ie when the Operator tries to reconcile Ingresses for the installation.

**Reply by grzesuav** [reply]
Hmm will look there. Is there any way to bypass it ? We do not use ingress, I just need native kubernetes service as we do not have ingress controller. We use istio virtual service

**Reply by andrewazores** [reply]
I don't think there is any way to bypass it right now - the Operator does assume that it will manage either Ingresses or Routes for the installation. Your manifest/argocd/istio setup is unfortunately pretty far outside of how we designed the Operator to be installed and used, so some of the assumptions we made don't hold in your setup :-(

**Reply by grzesuav** [reply]
I might create an issue to maybe revisit the assumptions or make it more flexible - I do not belive my setup is different than actual industry standard in 2024

**Reply by andrewazores** [reply]
That's fair. These assumptions were made some time around 5 years ago, and despite our efforts to be distribution-agnostic our actual development and testing mostly happens on OpenShift so our thinking will tend to be biased in that direction just because it's what we are most accustomed to and have available. I don't know that we will have much bandwidth to tackle these low-level assumptions in the Operator any time soon, however.

**Comment by grzesuav** [comment]
One more thing, I just noticed that after capturing a dump I have some copy - any idea how it got there ?
<img width="1332" alt="image" src="https://github.com/user-attachments/assets/29c108dc-b663-4b89-8273-da0031725146" />

**Reply by grzesuav** [reply]
got second sample - seems it happening when recording supposed to stop ? 
<img width="1641" alt="image" src="https://github.com/user-attachments/assets/cf2528ca-ec9c-49f3-8898-2beaff08b883" />

**Reply by andrewazores** [reply]
I have seen things like this before in 3.0.0/3.0.1, it seems like some bug that was introduced in the major rewrite at that version. I assume you have 3.0.1 deployed here?

There has been some recent work in this area which I think addressed the usual root causes of this bug (https://github.com/cryostatio/cryostat/pull/689), but there are a few more things I can think of trying.

**Reply by andrewazores** [reply]
I filed #760 to track this particular bug.

**Comment by grzesuav** [comment]
The last thing which seems suspicious (I did not have it on helm chart version) is 
<img width="1641" alt="image" src="https://github.com/user-attachments/assets/21325bee-fc15-4049-a894-815bfc085672" />
not really sure what it is about (not the grafana one, but `undefined` exception)

**Reply by andrewazores** [reply]
Hard to say what exactly is going wrong here. It looks as if the WebSocket connection from the client (your browser) to the Cryostat server instance is failing. Is there anything like a proxy in between there, and if so does it allow WebSocket connections?
